content/docs/rfcs/13/README.md

```diff
@@ -158,10 +158,11 @@ However, this `cursor` reusability  does not hold when the `receiverTime` is uti
 RPC call to respond to a HistoryQuery call.
 - The `messages` field MUST contain the messages found, these are [`WakuMessage`] types as defined in the corresponding [specification](./waku-message.md).
 - `PagingInfo`  holds the paging information based on which the querying node can resume its further history queries. 
-  The `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in the `messages` field of `HistoryResponse`). 
+  The `pageSize` indicates the number of returned Waku messages (i.e., the number of messages included in the `messages` field of `HistoryResponse`). 
   The `direction` is the same direction as in the corresponding `HistoryQuery`. 
   In the forward pagination, the `cursor` holds the `Index` of the last message in the `HistoryResponse` `messages` (and the first message in the backward paging). 
-  The requester shall embed the returned  `cursor` inside its next `HistoryQuery` to retrieve the next page of the waku messages.  
+  Regardless of the paging direction, the retrieved `messages` are always sorted in ascending order based on their timestamp, that is, from the oldest to the most recent.
```

Comment:
> Regardless of the paging direction, the retrieved `messages` are always sorted in ascending order based **on their timestamp**, that is, from the oldest to the most recent.

As far as I have understood, the message `timestamp` field (aka `senderTimestamp` field) is optional. What happens in the case the field is not present? Based on which criteria should you return the messages ordered?

IMO, and this is related to the issues that we are facing in the nwaku implementation:

* A store timestamp should be defined to have a unique timestamp for store operations.
* If the sender timestamp is provided, and it is within a window of [t-10s, t+10s], the message should be stored with the sender timestamp as the `storedAt` timestamp.
* If no sender timestamp is provided, we should set the current time as the `storedAt` timestamp.

Then all queries and message ordering should be based on that `storedAt` timestamp.

Reply:
The protobuf specs do not allow for a field to be `optional` per se. it's either set and interpreted or not set and interpret as default, which is `0` for number fields.
From a receiver POV, it is not possible to differentiate a sender that did not include the `timestamp` from a sender that set the `timestamp` to `0`

Hence, a message with no `timestamp` should not be present in the DB and I guess is likely to be purged whenever the retention policy is ran.

So, should a `timestamp` of `0` be considered like an `ephemeral=true` message and simply never be stored?

Reply:
Some thoughts on this @staheri14 @rymnc?

Should we extend the RFC to clarify that a message with a `timestamp` equal to zero SHOULD be considered as an ephemeral message for store purposes? I think it would help remove any implementation ambiguity.

Reply:
I think due to anonymity guarantees, we will eventually have no sender timestamp in the WakuMessage envelope. This means that we have to fall back to another ordering mechanism, which could be ordering based on when the receiver sees the message as @LNSD mentions. Adding to that proposal, I think we will have to mention that there are no *strong* ordering guarantees, because 2 nodes separated by a large distance geographically will see messages at different times, and therefore may not have the same order in their store.

Reply:
Below is my suggestion (part of it was discussed during the last PM call)
- Let's keep the sender time stamp but clarify the anonymity and consistent data availability trade-off 
- In case the timestamp is not present, let's fall back to the receiver time stamp (i.e., `storedAt`). The sender timestamp will not be altered, and the receiver timestamp is only used for sake of message ordering and is an internal logic. When a message with zero ts is retrieved in a history query, its timestamp will still be zero, but its timestamp can be approximated by looking at its adjacent messages. For example, if m0 is a message with zero timestamps, and a forward history query returns `x`, `m0`, `y` then it means from the storage node pov, x.ts < m0.ts < y.ts (ts: timestamp). 
- The above strategy is applicable to the `resume` procedure i.e., `FT-Store` with a slight modification. When an offline store node fetches a list of messages from another store node via time range query, it will rely on the ordering of the queried node to determine the ts of messages with no sender ts. That is if the query result is `x`, `m0`, `y`, where `x` and `y` have ts but not `m0`, then the querying store node can set an arbitrary receiver time for `m0` by e.g., averaging `y` and `x` timestamp, to other strategies.

Reply:
Sounds good. maybe worse specifying what `timestamp` means in the context of this change (ie, `timestamp` field or `receivedTimestamp` if the former is zero'd).

Reply:
> In case the timestamp is not present, let's fall back to the receiver time stamp (i.e., storedAt). The sender timestamp will not be altered, and the receiver timestamp is only used for sake of message ordering and is an internal logic.

Agree. I implement it like that in [#1178](https://github.com/status-im/nwaku/pull/1178).

With that, we have defined behavior for the "message validation" and "store ordering" problems.

> When a message with zero ts is retrieved in a history query, its timestamp will still be zero, but its timestamp can be approximated by looking at its adjacent messages. For example, if m0 is a message with zero timestamps, and a forward history query returns `x`, `m0`, `y`, then it means from the storage node pov, `x.ts < m0.ts < y.ts` (`ts`: timestamp).
>
> The above strategy applies to the `resume` procedure i.e., `FT-Store` with a slight modification. When an offline store node fetches a list of messages from another store node via time range query, it will rely on the ordering of the queried node to determine the ts of messages with no sender ts. That is if the query result is `x`, `m0`, `y`, where `x` and `y` have ts but not `m0`, then the querying store node can set an arbitrary receiver time for `m0` by e.g., averaging `y` and `x` timestamp, to other strategies.

Yes, that applies the particular use case of the resume store mechanism. I see this as a separate piece: _How do you interpret the "no-sender-timestamp" messages when retrieved from another store_. And as I am a divide and conquer guy, I would like to have this in a separate discussion/piece of work 😁

But, before that problem, we have to solve the "message duplication" criteria (that very likely will also apply to the "resume store" mechanism), or: _How do we detect duplicated messages when inserting those messages in the store?_

Basically, the question is: **When should consider that two messages are the same?**

Reply:
> Basically, the question is: When should consider that two messages are the same? 

Before diving into the answer to this, it would be good to first clarify whether the de-duplication is a store protocol concern or the relay protocol?

Reply:
Ok, based on your comment, under only one circumstance de-duplication is need and that is in the `resume` procedure, right?

Reply:
Given that we allow omitting sender timestamp for privacy's sake, then we cannot deduplicate such messages when transferring them between two different store nodes.  This follows our initial design as well where for a single store node, two identical messages whose timestamps are zero will be treated as two different messages. I'd say that should also hold for time-range queries. 

One solution for this issue can be adding a `nonce` field to the waku messages which is set completely at random by the message publisher. This would allow the deduplication of messages with zero timestamps because their `nonce`s will be disseminating. 

In general, I suspect we encounter such a duplication issue, because if an application cares about user privacy to the extent that omits the waku message sender timestamp, then supposedly it uses encryption for the `payload` field (otherwise it would not be privacy-preserving). Since a secure (CPA) encryption is a randomized algorithm, then two messages with the same plain payload (published and encrypted separately) will never have the same encrypted value, hence there will be no duplicate. Considering this, it would be very unlikely to get a  duplication issue.

Reply:
I agree that the most consistent solution for now is what Lorenzo implemented here: https://github.com/status-im/nwaku/pull/1178, namely using a single `storeTimestamp` for message ordering (and indexing) which is equal to the `senderTimestamp` (unless `senderTimestamp` is `0` and we fall back to a receiver generated timestamp).

> Basically, the question is: When should consider that two messages are the same?

I think we're on the same page based on the solution above, but to have it explicitly set out here. Two messages are the same when they have the same:
`timestamp`, `payload`, `contentTopic` and `pubsubTopic`. The exception here is indeed when both messages have no `timestamp` (i.e. `timestamp` is `0`) and we essentially only compare the remaining three fields (by falling back to `receiverTimestamp`, which will more or less always differ between subsequent messages).

> Before diving into the answer to this, it would be good to first clarify whether the de-duplication is a store protocol concern or the relay protocol?

Indeed deduplication should be performed by `relay`. However, defining a minimal set of uniqueness is an essential part of data model design, defining indexes and finding the proper ordering within the store protocol. This is necessary to build performant queries. One other subtle difference between `relay` and `store` is that `relay` always functions within a single `pubsubTopic`, whereas `store` can span multiple `pubsubTopic`s (although including `pubsubTopic` in the index implies that `store` will also treat them separately).

> only one circumstance de-duplication is needed and that is in the resume procedure, right?

I already think this new solution works well even for `resume()` in 99% of cases and will be only slightly inconsistent in the 1%. IMO we can identify any issues with this and address them separately once we have more formal requirements of the FT-STORE.

> One solution for this issue can be adding a nonce field to the waku messages

Good idea! One question would be how Waku users would be incentivised to use this field: I can see arguments for using a timestamp (proper consistent store ordering and retrieval) and against using a timestamp (better anonymity), but I'm not sure why the latter group would opt to generate a nonce instead, unless we make it a requirement for `store` validity. Definitely something to consider.

Reply:
> I already think this new solution works well even for resume() in 99% of cases and will be only slightly inconsistent in the 1%. IMO we can identify any issues with this and address them separately once we have more formal requirements of the FT-STORE.

If we exclude the cross-node queries i.e., `resume`, then I'd say that deduplication only makes sense for messages with sender timestamp. Without timestamps, we cannot deduplicate. For example, if a user sends a `payload=hello` message twice (under the same pubsub and content topic), two identical waku messages will be created (with no sender timestamp). When such messages arrive at a store node, they will be treated as two separate messages but not duplicates. The conclusion: 
- When the sender timestamp  is available -> deduplication is immediate and clear based on the timestamp
- When no sender timestamp is available -> deduplication is not applicable

Reply:
> Good idea! One question would be how Waku users would be incentivised to use this field: I can see arguments for using a timestamp (proper consistent store ordering and retrieval) and against using a timestamp (better anonymity), but I'm not sure why the latter group would opt to generate a nonce instead, unless we make it a requirement for store validity. Definitely something to consider.

"One question would be how Waku users would be incentivized to use this field" that is a good question. My suggestion was based on the assumption that there will be a significant number of duplicates due to the `resume` procedure. Otherwise, I do not think we would need to add nonce unless (practically) duplication becomes a real issue for messages with zero-timestamp.